Add WebGPU backend

spaarmann opened the pull request:
WIP WebGPU backend (#1403)

The JS glue generation is still very basic and barely enough to implement the server protocol on top of. There's also a slightly weird mixture of generating JS vs just accessing the embedded JSON manifest that I'm still hoping to clean up a bit.

spaarmann committed:
Remaining int and bool CmpOps in WGSL

spaarmann committed:
Factor out common call2 function

spaarmann committed:
More BinOps for WGSL

spaarmann committed:
Start supporting shift binops

spaarmann committed:
Finish shift binops

spaarmann committed:
Some initial plumbing bits for ad-hoc WGSL testing

spaarmann committed:
Support UnOpExps and fix i64 equality binop

spaarmann committed:
Merge remote-tracking branch 'origin/master' into webgpu

spaarmann committed:
Declare ConstUse variables as i64 but only pass lower 32 bits for now

spaarmann committed:
Define KernelM and mkGlobalIdent

spaarmann committed:
Start using KernelM

spaarmann committed:
Make module-scope names unique (generates invalid copies for buffers atm)

spaarmann committed:
Add renaming mechanism, rename memory buffers to be unique

spaarmann committed:
Continue on testing setup

spaarmann committed:
Format numbers as valid JS numbers in webgpu tests

spaarmann committed:
Give each kernel its own bind group

spaarmann committed:
Include kernel interface information in WebGPU test data output

spaarmann committed:
Generate override decls with default values

spaarmann committed:
Basic working JS WGSL primitive test runner

spaarmann committed:
Implement more i32 binops

spaarmann committed:
Improved test runner output

spaarmann committed:
Fix shift operators

spaarmann committed:
Generate all bindings into a single bind group

spaarmann committed:
Add pow i64 support

spaarmann committed:
Start on supporting i64 in WGSL tests

spaarmann committed:
Support i64 arrays in WGSL tests

spaarmann committed:
Fix bug in negate_i64 (and thus sub_i64)

spaarmann committed:
Fix generation of 64-bit integer literals

spaarmann committed:
Ability to run multiple test files, show results on page

spaarmann committed:
Merge remote-tracking branch 'origin/master' into webgpu

spaarmann committed:
Fix tab indentation in futhark.cabal

spaarmann committed:
Add testing support for u32, u64

spaarmann committed:
Fix comparison with expected values for 64-bit ints

spaarmann committed:
Add f32 testing support

spaarmann committed:
Enumerate unimplemented statements for better overview

spaarmann committed:
Fix some testing error handling

spaarmann committed:
Add some float built-in functions

spaarmann committed:
Merge remote-tracking branch 'origin/master' into webgpu

spaarmann committed:
Nicer test result layout

spaarmann committed:
Rename arith.wgsl to scalr.wgsl

spaarmann committed:
Start on i8 support

spaarmann committed:
More i8 progress

spaarmann committed:
Finish basic i8 support

spaarmann committed:
Basic support for booleans in storage buffers

spaarmann committed:
bool output support in tests; improve robustness of finding right kernel

spaarmann committed:
Print expected output according to its type rather than input type

spaarmann committed:
Initial i16 support

spaarmann committed:
bool <-> integer conversions

spaarmann committed:
Some tolerance for float comparisons for primitive tests

spaarmann committed:
Start implementing GPU abstraction layer for WebGPU

spaarmann committed:
Rest of the abstraction layer other than kernel creation and launching

spaarmann committed:
Merge remote-tracking branch 'origin/master' into webgpu

spaarmann committed:
Some WebGPU initialization code

spaarmann committed:
Re-use scalar readback buffer, some more initialization + teardown

spaarmann committed:
Initial gpu_create_kernel and gpu_launch_kernel draft

spaarmann committed:
Add uncaptured error callback

spaarmann committed:
Start on Haskell side of WebGPU C code generation

spaarmann committed:
Factor out common host code definitions into new module so that WebGPU
backend can use CodeGen.Backends.GPU

spaarmann committed:
Refactor KernelInterface stuff a bit, compute proper scalars struct layout

spaarmann committed:
Generate static wgpu_kernel_info array

spaarmann committed:
Add webgpu CLI command, fix a bunch of errors in webgpu.h

spaarmann committed:
More RTS and code gen fixes, enough to compile a whole program to WASM

spaarmann committed:
Generate MacroDefs for ConstUses and builtin overrides

spaarmann committed:
Fix some more errors, actually get a compiling program

spaarmann committed:
Actually pass arguments in LaunchKernel

spaarmann committed:
Some temporary debug logging

spaarmann committed:
Fix kernel info for launches not being found, remove debug logs

spaarmann committed:
Experimental support for compiling against Dawn

spaarmann committed:
Merge remote-tracking branch 'origin/master' into webgpu

spaarmann committed:
Mark entry points to be exported

spaarmann committed:
Export a very basic JS interface

spaarmann committed:
Start on `futhark test` server wrapper

spaarmann committed:
Start on generating a proper JS interface

spaarmann committed:
Small fixes

spaarmann committed:
More JS wrapper stuff

spaarmann committed:
Fix JS interface, add some debugging infrastructure stuff

spaarmann committed:
Small bits on test server

spaarmann committed:
Don't generate semicolons at the end of WGSL struct decls

spaarmann committed:
Basic WS server proxying structure

spaarmann committed:
First server protocol command

spaarmann committed:
inputs and outputs commands, start on restore

spaarmann committed:
Enough commands and fixes for `futhark test tests/soacs/map2.fut` to pass

spaarmann committed:
More reasonable CLI arg parsing in Python runner

spaarmann committed:
Very basic naive way of spawning a headless chrome process to run WebGPU tests

spaarmann committed:
Properly use Selenium for driving browser

spaarmann committed:
Remove __pycache__ files

spaarmann committed:
Remove some debug and old stuff

spaarmann committed:
Merge remote-tracking branch 'origin/master' into webgpu

spaarmann committed:
Work around Chrome/Dawn bug by only passing override consts used by the entry point

spaarmann committed:
Support for futhark bench

spaarmann committed:
Improve benchmarking timer accuracy

spaarmann committed:
Default chrome to start headless

spaarmann committed:
Ignore generates .js files in tests dir

spaarmann committed:
Ignore python cache in rts/python

spaarmann committed:
Couple of fixes + more options for browser_test runner

spaarmann committed:
Support more types for WASM<->JS interactions, enable WASM_BIGINT

athas commented on rts/python/browser_test.py:
This should go in `tools/`.

athas commented on rts/python/browser_test.py:
```diff
@@ -0,0 +1,303 @@
+#!/usr/bin/env python
```
`python3` surely?

athas commented on rts/wgsl/scalar.wgsl:
For consistency, I suggest moving the 32-bit operations into a file `scalar32.wgsl`.

athas commented on src/Futhark/Actions.hs:
```diff
@@ -224,6 +248,7 @@ runCC cpath outpath cflags_def ldflags = do
             ++ cmdCFLAGS cflags_def
             ++
             -- The default LDFLAGS are always added.
+            -- The default LDFLAGS are always added.
```
Duplicate line.

athas commented on src/Futhark/Actions.hs:
```diff
@@ -260,6 +285,8 @@ runISPC ispcpath outpath cpath ispcextension ispc_flags cflags_def ldflags = do
             ++ cmdCFLAGS cflags_def
             ++
             -- The default LDFLAGS are always added.
+            -- The default LDFLAGS are always added.
+            -- The default LDFLAGS are always added.
```
Two duplicate lines.

athas commented on src/Futhark/CodeGen/Backends/CWebGPU.hs:
```diff
@@ -0,0 +1,382 @@
+       #endif
+      |]
+
+-- | As server script. Speaks custom server protocol to local Python server
```
I suggest not using the term "server protocol" to reference anything but the "Futhark server protocol" (which should be understood as "(Futhark server) protocol". The term "server protocol" on its own is either wrong or redundant. Just call a  "custom protocol" or "browser test protocol" if it must have a name.

athas committed:
Factor WGSL definition into separate package.

spaarmann committed:
Merge pull request #2145 from diku-dk/webgpu-package

Factor WGSL definition into separate package.

spaarmann commented on tools/extract-shader.sh:
No, this was for very early testing, I'll remove it.

spaarmann committed:
Address some feedback

athas committed:
Add more browser crud.

athas commented:
I pushed a commit that enables some more browser crud. With that change, it works on my home machine. (Probably not on my laptop, which I think has a GPU that is blocklisted by Chromium.)

spaarmann committed:
Don't set Linux-specific options when using remote chrome driver

spaarmann committed:
Revamp JS interface to rely less on codegen and be more usable; do binary data format parsing and reading in JS instead of python wrapper

spaarmann committed:
Move browser_test.py to tools/

spaarmann committed:
Move i32/u32 specific WGSL functions into scalar32.wgsl

spaarmann commented:
> I pushed a commit that enables some more browser crud.
I guarded the extra arguments to not apply when using a remote driver as heuristic for my WSL setup since it seems passing them to Windows Chrome is not allowed. I assume this is fine for your home machine, but we might need a better chance if we ever use the remote mechanism for anything else.

I also revamped the JS interface to be much more usable. I ended up making a new JS implementation of the binary data protocol. There were some problems with using the one for the WASM backend, it uses some node-specific things and isn't exactly set up to deal with some of the API functions suddenly being asynchronous. Didn't want to mess with it too much without having looked much at the WASM backend codegen etc. and the format is simple enough.

spaarmann committed:
Round up host2gpu memcpy to multiple of 4

spaarmann committed:
Same rounding for gpu_memcpy

spaarmann committed:
Implement division ops for {i,u}{8,16}

spaarmann committed:
Support passing full 64-bit ConstUses

spaarmann committed:
Add (not very well tested yet) support for dynamic and >1d block dims

spaarmann committed:
Handle kernels without any ScalarUses

spaarmann committed:
Some ormolu formatting

spaarmann committed:
More ormolu formatting

spaarmann committed:
Python formatting

spaarmann committed:
Make tools/browser_test.py type-check

spaarmann committed:
Rework monad structure in ImpGen.WebGPU a bit to make adding module-declarations while translating kernel body more feasible

spaarmann committed:
Add support for some shared memory

spaarmann committed:
Limited i64 division ops, reduce duplication in `wgslBinOp`

spaarmann committed:
Merge remote-tracking branch 'origin/master' into webgpu

spaarmann committed:
Merge remote-tracking branch 'origin/master' into webgpu

spaarmann committed:
Handle booolean scalars properly

spaarmann committed:
signed int to f32 conversions

spaarmann committed:
Merge remote-tracking branch 'origin/master' into webgpu

spaarmann committed:
Fix literal 1 with 64-bit loop counters

spaarmann committed:
Fix findMemoryTypes not traversing through code nested in comments

spaarmann committed:
Always enable f16 extension

spaarmann committed:
int->float conversions

spaarmann committed:
Start on supporting atomics

spaarmann committed:
Merge remote-tracking branch 'origin/master' into webgpu

spaarmann committed:
wip reading small integer shared memory

spaarmann committed:
Work around chrome bug when lockstep width is not accessed

spaarmann committed:
Use normal types for shared memory instead of atomic<i32> for small types

spaarmann committed:
Merge remote-tracking branch 'origin/master' into webgpu

spaarmann committed:
Prototype of UniformRead KernelOp

spaarmann committed:
Implement UniformRead as a normal read on other GPU backends

spaarmann committed:
Some float to int conversions

spaarmann committed:
Remaining atomic ops

spaarmann committed:
Fix WGSL bool literal pretty printing

spaarmann committed:
rts: define max_thread_block_size

spaarmann committed:
rts: Check if copy padding could corrupt data

spaarmann committed:
Also generate barriers for ErrorSyncs

spaarmann committed:
Add some more sizes

spaarmann committed:
atomicStore/Load for Write/Reads from atomically-typed buffers

spaarmann committed:
JS interface: `entry` object instead of `entry_` prefix

athas committed:
Loosen bound.

spaarmann committed:
Compile with optimizations instead of debug info

spaarmann committed:
Merge remote-tracking branch 'origin/master' into webgpu

spaarmann committed:
Handle `AtomicWrite`s

spaarmann committed:
More robust base64 encoding in JS

spaarmann committed:
f{16,32} to signed 8,16,32 bit integers

athas committed:
Merge branch 'master' into webgpu

athas committed:
Apparently renamed.

athas committed:
Whitespace cleanup.

athas committed:
Constness fix.

athas committed:
Merge branch 'master' into webgpu

athas committed:
Merge branch 'master' into webgpu

candrdk committed:
fix webgpu compile errors

athas committed:
Merge branch 'master' into caleb-webgpu

athas committed:
Bump Nixpkgs for WebGPU.

athas committed:
Necessary cast.

athas committed:
Add handy environment variable.

athas committed:
Remove some crap.

candrdk committed:
Handle infinities and NaN constants for f32 and f16 on WebGPU

candrdk committed:
start work on adding builtin kernels to webgpu backend

candrdk committed:
Implement builtin copy kernel. Start work on generation of 1b, 2b, 4b and 8b variations of builtin transpose/copy kernels

candrdk committed:
fix up support for different elem types in builtin_kernels.

candrdk committed:
rebuild on builtin_kernels.wgsl changes

athas committed:
Merge branch 'master' into webgpu

athas committed:
Enforce style.

candrdk committed:
Split builtin kernels into seperate shader modules. Start work on optimized kernels.

candrdk committed:
delete old builtin_kernels.wgsl

candrdk committed:
Use the appropriate shadermodule when launching kernel.

candrdk committed:
optimized builtin kernels

candrdk committed:
fix up small kernel. Still need to test the rest

candrdk committed:
fix bug

candrdk committed:
Initial support for f32 atomics

candrdk committed:
Make sure dims are all actually 64-bit expressions. While not a problem for other backends, WebGPU currently can't handle multiplication between differing int sizes

candrdk committed:
Continue work on 8/16-bit atomics. Just using atomicLoad/Store as placeholders for now, have to implement proper CAS later

candrdk committed:
More atomics work. Still trying to debug gpu hung error when writing with atomicCompareExchangeWeak

candrdk committed:
atomic read fix

candrdk committed:
Get i8 and i16 atomics working. Still missing f16. Also, some atomic ops (xchg, cmpxchg) still only support i32. Some atomic ops haven't been tested yet

athas committed:
Merge branch 'master' into webgpu

athas committed:
Makefile: add test-webgpu

candrdk committed:
atomics cleanup

candrdk committed:
fix multiline comments

candrdk committed:
Increase max msg size over websocket. Default limit is very small, resulting in hangs with non-trivial input/output array sizes

candrdk committed:
enable bool atomics, improve error message

athas committed:
Merge branch 'master' into webgpu

athas committed:
Merge branch 'master' into webgpu

athas committed:
Support 16 bit atomics in OpenCL.

candrdk committed:
Add support for in-kernel copies

candrdk committed:
Merge branch 'diku-dk:webgpu' into webgpu

candrdk committed:
Proper support for arr size dimensioning of ScalarSpace mem declarations

candrdk committed:
Improve primitives coverage. Fix negation of integers bug. Make setmem statements an error.

athas committed:
Merge branch 'master' into webgpu

candrdk committed:
WIP: Debugging buffer memmoves

candrdk committed:
Add f16 atomics

candrdk committed:
atomics cleanup

candrdk committed:
Generate device functions, allowing them to be called from kernels

candrdk committed:
Improve primitive test coverage

candrdk committed:
Small additions/fixes

candrdk committed:
clean up memmove

athas committed:
Merge branch 'master' into webgpu

athas committed:
Merge remote-tracking branch 'candrdk/webgpu' into webgpu

athas committed:
Nicer errors.

athas committed:
Style fixes.

athas committed:
Clearer error.

athas committed:
Temporary inlining hack.

candrdk committed:
fix wgsl nan/inf primitives. Add naninf16 test

candrdk committed:
Merge branch 'webgpu' of github.com:candrdk/futhark into webgpu

candrdk committed:
atomics cleanup

candrdk committed:
Fix up signed/unsigned atomics. We just declare all as i32 and cast as necessary. This does mean we end up simulating 32-bit max/min atomic ops with CAS loops instead of using the native wgsl versions. Worth the tradeoff for better compativbility, however.

candrdk committed:
Use max possible limits. Since we cannot know the bound storage buffer limit up-front, warning has just been removed for now.

candrdk committed:
fix min/max tests

candrdk committed:
Support function calls with multiple return parameters

candrdk committed:
fix

candrdk committed:
forgot to push

candrdk committed:
Fix single-return calls

candrdk committed:
fix for real this time

candrdk committed:
and atomic still doesn't work for some reason...

candrdk committed:
clean up transpose kernels

candrdk committed:
fix 64-bit integer literals

candrdk committed:
Merge branch 'webgpu' of github.com:candrdk/futhark into webgpu

candrdk committed:
Take memory space into account when generating reads and writes. If reading from a local array in ScalarSpace, for example, 8- and 16-bit values are padded and should be accessed through direct indexing, instead of read_*/write_* functions

candrdk committed:
Fix out-of-memory crashes caused by large test inputs

candrdk committed:
fix atomic_and16

athas committed:
Merge branch 'master' into webgpu

athas committed:
Merge branch 'master' into webgpu

athas committed:
Merge branch 'master' into webgpu

athas committed:
This should not exist anymore.

athas committed:
More tuning parameter fixes.

athas committed:
Remove unused import.

Jesssullivan committed:
Update WebGPU emdawnwebgpu compatability, add many tests 🖖  (#2365)

Co-authored-by: Troels Henriksen <athas@sigkill.dk>
Co-authored-by: Tudor Pal <pal.tudor10@gmail.com>

athas committed:
Merge branch 'master' into webgpu

athas committed:
Hacky fix for opaques.

athas committed:
Add wrapper class for opaques.

Closes #2368.

athas committed:
Merge branch 'master' into webgpu

TudorPal committed:
webgpu hendrix test installation workflow

TudorPal committed:
WebGPU test without the futhark-slurm action

athas committed:
Merge branch 'master' into webgpu

athas committed:
Reformat a bit.

athas committed:
Add some final linebreaks.

TudorPal committed:
move the webgpu CI to main, update setup scripts to cache installations

TudorPal committed:
temporarily add webgpu to the branch list to test the main workflow update

TudorPal committed:
webgpu CI looks fine, removing it from the push branches list

teddyjfpender commented:
Do you guys need any help moving this forward? Happy to help :)

athas committed:
Merge branch 'master' into webgpu

athas commented:
You are very welcome to help out. Currently matters are a bit unstructured, I'm afraid.

* @TudorPal has been working on improving the documentation and testing situation, and in particular has developed some examples that are available here: https://github.com/diku-dk/futhark-webgpu
* A very useful contribution would be to look at those examples and see whether they make sense for you.
* Then, I suggest trying to use the backend for something, and figure out where it falls short.

WebGPU has limitations that mean some things simply cannot work. Eventually we should have good error messages (or maybe fallbacks) for those, but we don't have a good list of the current limitations.

athas committed:
Remember this dependency.

athas committed:
Style fix.

athas committed:
Restore wasm backend.

TudorPal committed:
Reconcile WebGPU and WASM JavaScript APIs, as per diku-dk/futhark-webgpu#8 (#2475)

* Align WebGPU JS wrapper with shared API

Update the WebGPU JavaScript wrapper to better match the WASM wrapper API.
Arrays now expose shape() as the shared shape accessor, with the internal
shape field renamed to avoid the name clash. Add WASM-style convenience
methods to WebGPU arrays, including toTypedArray(), toArray(), and
futharkType().

Also update entry point wrappers so single-output entry points return the
single value directly, matching the WASM API, while multi-output entry
points still return an array. Add WebGPU API tests for scalar entry calls,
array construction, returned arrays, and the shared FutharkModule interface.

* adding the tunnel example from https://github.com/diku-dk/futhark-webgpu

* inc test was only for headless browser, added another one to be able to test on local browser to test the full runtime behaviour manually

* array browser test added

athas committed:
Merge branch 'master' into webgpu
